File: frontend/src/types/requisition.ts
export type QuantityMode = "numeric" | "fulltank";

export interface Requisition {
  id: number;
  serial_number?: string | null;
  quantity: number;
  unit: string;
  unit_price?: number | null;
  total_price?: number | null;
  status: string;
  notes?: string | null;
  fuel_type?: string | null;
  requestor_name?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  is_edited: boolean;
  edited_by?: string | null;
  requester_id?: number | null;
  vehicle_id?: number | null;
  vendor_id?: number | null;
  plate_number?: string | null;
  model?: string | null;
  vendor_name?: string | null;
  can_edit: boolean;
  edit_error?: string | null;
}

export interface PendingRequisition extends Requisition {
  requester_name?: string | null;
  display_total?: number | null;
  can_approve: boolean;
}

export interface RequestFormContext {
  default_requestor_name: string;
  vehicles: FormVehicle[];
  vendors: FormVendor[];
}

export interface FormVehicle {
  id: number;
  plate_number: string;
  model: string;
  vendor_id?: number | null;
  vendor_name?: string | null;
  fuel_type?: string | null;
  company?: string | null;
}

export interface FormVendor {
  id: number;
  name: string;
  address?: string | null;
}

export interface RequisitionCreateRequest {
  vehicle_id: number;
  vendor_id: number;
  fuel_type: string;
  quantity_mode: QuantityMode;
  quantity: number;
  notes?: string;
  requestor_name: string;
}

export interface RequisitionUpdateRequest {
  vehicle_id: number;
  vendor_id: number;
  fuel_type: string;
  quantity_mode: QuantityMode;
  quantity: number;
  notes?: string;
}

export interface ApprovalQuantityUpdateRequest {
  quantity_mode: QuantityMode;
  quantity: number;
  notes?: string;
}

export interface PriorApproved {
  vehicle_id: number;
  requests: Array<Record<string, unknown>>;
}

File: api/schemas/requisition.py
"""Requisition API schemas."""
from typing import Literal, Optional

from pydantic import BaseModel, Field

QuantityMode = Literal["numeric", "fulltank"]


class RequisitionResponse(BaseModel):
    """Fuel requisition record for API responses."""

    id: int
    serial_number: Optional[str] = None
    quantity: float
    unit: str
    unit_price: Optional[float] = None
    total_price: Optional[float] = None
    status: str
    notes: Optional[str] = None
    fuel_type: Optional[str] = None
    requestor_name: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    is_edited: bool = False
    edited_by: Optional[str] = None
    requester_id: Optional[int] = None
    vehicle_id: Optional[int] = None
    vendor_id: Optional[int] = None
    plate_number: Optional[str] = None
    model: Optional[str] = None
    vendor_name: Optional[str] = None
    can_edit: bool = False
    edit_error: Optional[str] = None


class RequisitionCreateRequest(BaseModel):
    """Payload to submit a new fuel requisition."""

    vehicle_id: int
    vendor_id: int
    fuel_type: str = Field(..., min_length=1)
    quantity_mode: QuantityMode
    quantity: float = 0.0
    notes: str = ""
    requestor_name: str = Field(..., min_length=1)


class RequisitionUpdateRequest(BaseModel):
    """Payload to update a pending requisition."""

    vehicle_id: int
    vendor_id: int
    fuel_type: str = Field(..., min_length=1)
    quantity_mode: QuantityMode
    quantity: float = 0.0
    notes: str = ""


class ApprovalQuantityUpdateRequest(BaseModel):
    """Payload for approvers to edit quantity before approval."""

    quantity_mode: QuantityMode
    quantity: float = 0.0
    notes: str = ""


class PendingRequisitionResponse(RequisitionResponse):
    """Pending requisition with approval display fields."""

    requester_name: Optional[str] = None
    display_total: Optional[float] = None
    can_approve: bool = False


class RequestFormContextResponse(BaseModel):
    """Context data for the new request form."""

    default_requestor_name: str
    vehicles: list
    vendors: list


class PriorApprovedResponse(BaseModel):
    """Prior approved requests warning for a vehicle."""

    vehicle_id: int
    requests: list
